### PR TITLE
Add support for Netty-Server's UnixChannelOption.SO_REUSEPORT

### PR DESCRIPTION
**Pull request by fogan**
Two feature were added in this PR
1.Port reuse,when using ShadedNettyGrpcServerFactory,anather process can start its server in the same port. 
2.Graceful termination,when receive the signal of kill -15，the service does not stop immediately, but first completes the current task and then exits.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -9,7 +9,7 @@ buildscript {
         }
     }
     ext {
-        projectVersion = '2.10.1.RELEASE'
+        projectVersion = '2.10.1.1.RELEASE'
```
Comment (ST-DDT) [inline build.gradle:12]: Please don't change the version number arbitarily in the PR. If you need a fast release, mention this in the description and we try to do that.

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/serverfactory/ShadedNettyGrpcServerFactory.java [hunk 2]
```diff
@@ -63,9 +64,23 @@ protected NettyServerBuilder newServerBuilder() {
...
         } else {
-            return NettyServerBuilder.forAddress(new InetSocketAddress(InetAddresses.forString(address), port));
+            NettyServerBuilder nettyServerBuilder=NettyServerBuilder.forAddress(new InetSocketAddress(InetAddresses.forString(address), port));
+            if(this.properties.isReUsePort()){
+                nettyServerBuilder.withOption(UnixChannelOption.SO_REUSEPORT,true);
+            }
+            if(this.properties.isReUseAddr()){
+                nettyServerBuilder.withOption(UnixChannelOption.SO_REUSEADDR,true);
+            }
+            return nettyServerBuilder;
```
Comment (ST-DDT) [inline grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/serverfactory/ShadedNettyGrpcServerFactory.java:83]: Please also implement it for the non-shaded grpc server factory.

**Review by ST-DDT — CHANGES_REQUESTED** [review]
I'm not familiar with these options, the netty javadocs were empty and all I could find was "helps with faster restarts". How does it work/What does it do?

Please update your PR to the latest changes to resolve the merge conflicts.

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/serverfactory/ShadedNettyGrpcServerFactory.java (hunk 2 above)
Comment (fogan) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/serverfactory/ShadedNettyGrpcServerFactory.java:83]: > Please also implement it for the non-shaded grpc server factory.

This feature in shaded  is supported，but non-shaded is  not supported，which is given by grpc-community。
can u help me pr，and give me a fast verison or  snapshot，i am waitting to use this feature in my project。

Comment (ST-DDT) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/serverfactory/ShadedNettyGrpcServerFactory.java:83]: > > Please also implement it for the non-shaded grpc server factory.
> 
> This feature in shaded is supported，but non-shaded is not supported，which is given by grpc-community。

AFAIK grpc-netty and grpc-shaded-netty ues the exact same code, just with different packages.
I'll probably try it out myself.
Can you give me a little hint what this SO_REUSEPORT/SO_REUSEADDR does? Why does it need both?

> can u help me pr，and give me a fast verison or snapshot，i am waitting to use this feature in my project。

I could merge this faster, if you respond to my requests...
The v2.11.0 release will be ready/deployed soon, with or without this feature.
You can always build the jars locally. `gradlew publishToMavenLocal`

**Comment by ST-DDT** [comment]
Closing because the questions related to this feature proposal haven't been answered at all.

**Pull request closed without merging**